core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -408,13 +408,31 @@ public void index(Index index) {
                     innerIndex(index);
                 }
             }
-        } catch (IllegalStateException | IOException e) {
-            try {
-                maybeFailEngine("index", e);
-            } catch (Exception inner) {
-                e.addSuppressed(inner);
+        } catch (Exception e) {
+            handleOperationFailure(index, e);
+        }
+    }
+
+    /**
+     * When indexing a document into Lucene, Lucene distinguishes between environment related errors
+     * (like out of disk space) and document specific errors (like analysis chain problems) by setting
+     * the IndexWriter.getTragicEvent() value for the former. maybeFailEngine checks for these kind of
+     * errors and returns true if that is the case. We use that to indicate a document level failure
+     * and set the error in operation.setFailure. In case of environment related errors, the failure
+     * is bubbled up
+     */
+    private void handleOperationFailure(final Operation operation, final Exception e) throws OperationFailedEngineException {
```

Comment:
I restructured the method to clarify what we are doing here. We only wrap failures that fail the engine for serialization. Any request/document specific failures are communicated to the transport via `Operation#setFailure` and `Operation#getFailure`.